File: backend/app/services/analytics.py
from typing import List, Dict, Optional
from datetime import datetime, timedelta
from sqlalchemy import select, func, desc
from sqlalchemy.ext.asyncio import AsyncSession
from app.models.test import TypingTest
from app.models.analytics import UserAnalytics
from app.models.user import User
from uuid import UUID


class AnalyticsService:
    async def get_user_analytics(self, db: AsyncSession, user_id: UUID) -> Optional[Dict]:
        result = await db.execute(
            select(UserAnalytics).where(UserAnalytics.user_id == user_id)
        )
        analytics = result.scalar_one_or_none()
        if not analytics:
            return None
        return {
            "total_tests": analytics.total_tests,
            "total_time_seconds": analytics.total_time_seconds,
            "avg_wpm": analytics.avg_wpm,
            "avg_accuracy": analytics.avg_accuracy,
            "best_wpm": analytics.best_wpm,
            "best_accuracy": analytics.best_accuracy,
            "consistency_score": analytics.consistency_score,
            "weak_words": analytics.weak_words,
            "left_hand_error_rate": analytics.left_hand_error_rate,
            "right_hand_error_rate": analytics.right_hand_error_rate,
            "shift_key_error_rate": analytics.shift_key_error_rate,
            "number_row_error_rate": analytics.number_row_error_rate,
            "common_mistypes": analytics.common_mistypes,
            "fatigue_start_time": analytics.fatigue_start_time,
            "wpm_trend": analytics.wpm_trend,
            "accuracy_trend": analytics.accuracy_trend,
        }

    async def update_user_analytics(self, db: AsyncSession, user_id: UUID, test_result: TypingTest):
        result = await db.execute(
            select(UserAnalytics).where(UserAnalytics.user_id == user_id)
        )
        analytics = result.scalar_one_or_none()

        if not analytics:
            analytics = UserAnalytics(user_id=user_id)
            db.add(analytics)

        analytics.total_tests += 1
        analytics.total_time_seconds += int(test_result.time_taken_seconds or 0)

        if test_result.net_wpm:
            if analytics.avg_wpm:
                analytics.avg_wpm = (analytics.avg_wpm * (analytics.total_tests - 1) + test_result.net_wpm) / analytics.total_tests
            else:
                analytics.avg_wpm = test_result.net_wpm

        if test_result.accuracy:
            if analytics.avg_accuracy:
                analytics.avg_accuracy = (analytics.avg_accuracy * (analytics.total_tests - 1) + test_result.accuracy) / analytics.total_tests
            else:
                analytics.avg_accuracy = test_result.accuracy

        if test_result.net_wpm and (not analytics.best_wpm or test_result.net_wpm > analytics.best_wpm):
            analytics.best_wpm = test_result.net_wpm

        if test_result.accuracy and (not analytics.best_accuracy or test_result.accuracy > analytics.best_accuracy):
            analytics.best_accuracy = test_result.accuracy

        if test_result.consistency_score:
            if analytics.consistency_score:
                analytics.consistency_score = (analytics.consistency_score * (analytics.total_tests - 1) + test_result.consistency_score) / analytics.total_tests
            else:
                analytics.consistency_score = test_result.consistency_score

        if test_result.weak_words:
            existing_weak = analytics.weak_words or []
            new_weak = test_result.weak_words
            combined = existing_weak + new_weak
            word_counts = {}
            for w in combined:
                if isinstance(w, str):
                    word_counts[w] = word_counts.get(w, 0) + 1
            sorted_words = sorted(word_counts.items(), key=lambda x: -x[1])
            analytics.weak_words = [w for w, _ in sorted_words[:50]]

        last_20 = analytics.last_20_test_ids or []
        last_20.append(str(test_result.id))
        analytics.last_20_test_ids = last_20[-20:]

        wpm_trend = analytics.wpm_trend or []
        if test_result.net_wpm:
            wpm_trend.append({"test_id": str(test_result.id), "wpm": test_result.net_wpm, "date": datetime.utcnow().isoformat()})
            analytics.wpm_trend = wpm_trend[-50:]

        accuracy_trend = analytics.accuracy_trend or []
        if test_result.accuracy:
            accuracy_trend.append({"test_id": str(test_result.id), "accuracy": test_result.accuracy, "date": datetime.utcnow().isoformat()})
            analytics.accuracy_trend = accuracy_trend[-50:]

        await db.flush()

    async def get_leaderboard(
        self,
        db: AsyncSession,
        scope: str = "global",
        period: str = "all_time",
        limit: int = 100,
        user_id: Optional[UUID] = None,
        state: Optional[str] = None,
        district: Optional[str] = None,
        city: Optional[str] = None,
        college: Optional[str] = None,
    ) -> Dict:
        query = (
            select(User)
            .where(User.is_active == True)
            .order_by(User.xp.desc())
            .limit(limit)
        )

        if state:
            query = select(User).where(User.state == state, User.is_active == True).order_by(User.xp.desc()).limit(limit)
        if district:
            query = select(User).where(User.district == district, User.is_active == True).order_by(User.xp.desc()).limit(limit)
        if city:
            query = select(User).where(User.city == city, User.is_active == True).order_by(User.xp.desc()).limit(limit)
        if college:
            query = select(User).where(User.college == college, User.is_active == True).order_by(User.xp.desc()).limit(limit)

        result = await db.execute(query)
        users = result.scalars().all()

        entries = []
        user_rank = None
        for idx, user in enumerate(users):
            rank = idx + 1
            entries.append({
                "rank": rank,
                "user_id": str(user.id),
                "full_name": user.full_name,
                "college": user.college,
                "state": user.state,
                "city": user.city,
                "best_wpm": user.best_wpm or 0,
                "best_accuracy": user.best_accuracy or 0,
                "tests_taken": user.total_tests_taken,
                "xp": user.xp,
            })
            if user_id and user.id == user_id:
                user_rank = rank

        return {
            "entries": entries,
            "user_rank": user_rank,
            "total_users": len(users),
        }


analytics_service = AnalyticsService()

File: frontend/src/server/services/analytics.ts
import { db } from '../db/client';
import { userAnalytics } from '../db/schema/user-analytics';
import { errorPatterns } from '../db/schema/error-patterns';
import { users } from '../db/schema/users';
import { eq, and, desc } from 'drizzle-orm';

export class AnalyticsService {
  async getUserAnalytics(userId: string) {
    const [analytics] = await db
      .select()
      .from(userAnalytics)
      .where(eq(userAnalytics.userId, userId))
      .limit(1);

    if (!analytics) return null;

    return {
      totalTests: analytics.totalTests,
      totalTimeSeconds: analytics.totalTimeSeconds,
      avgWpm: analytics.avgWpm,
      avgAccuracy: analytics.avgAccuracy,
      bestWpm: analytics.bestWpm,
      bestAccuracy: analytics.bestAccuracy,
      consistencyScore: analytics.consistencyScore,
      weakWords: analytics.weakWords,
      leftHandErrorRate: analytics.leftHandErrorRate,
      rightHandErrorRate: analytics.rightHandErrorRate,
      shiftKeyErrorRate: analytics.shiftKeyErrorRate,
      numberRowErrorRate: analytics.numberRowErrorRate,
      commonMistypes: analytics.commonMistypes,
      fatigueStartTime: analytics.fatigueStartTime,
      wpmTrend: analytics.wpmTrend,
      accuracyTrend: analytics.accuracyTrend,
    };
  }

  async updateUserAnalytics(userId: string, testResult: Record<string, any>) {
    let [analytics] = await db
      .select()
      .from(userAnalytics)
      .where(eq(userAnalytics.userId, userId))
      .limit(1);

    if (!analytics) {
      [analytics] = await db
        .insert(userAnalytics)
        .values({
          userId,
          totalTests: 0,
          totalTimeSeconds: 0,
        })
        .returning();
    }

    const totalTests = analytics.totalTests + 1;
    const totalTimeSeconds = analytics.totalTimeSeconds + Math.round(testResult.time_taken_seconds ?? 0);

    let avgWpm = analytics.avgWpm;
    if (testResult.net_wpm != null) {
      avgWpm = avgWpm != null
        ? (avgWpm * (totalTests - 1) + testResult.net_wpm) / totalTests
        : testResult.net_wpm;
    }

    let avgAccuracy = analytics.avgAccuracy;
    if (testResult.accuracy != null) {
      avgAccuracy = avgAccuracy != null
        ? (avgAccuracy * (totalTests - 1) + testResult.accuracy) / totalTests
        : testResult.accuracy;
    }

    let bestWpm = analytics.bestWpm;
    if (testResult.net_wpm != null && (bestWpm == null || testResult.net_wpm > bestWpm)) {
      bestWpm = testResult.net_wpm;
    }

    let bestAccuracy = analytics.bestAccuracy;
    if (testResult.accuracy != null && (bestAccuracy == null || testResult.accuracy > bestAccuracy)) {
      bestAccuracy = testResult.accuracy;
    }

    let consistencyScore = analytics.consistencyScore;
    if (testResult.consistency_score != null) {
      consistencyScore = consistencyScore != null
        ? (consistencyScore * (totalTests - 1) + testResult.consistency_score) / totalTests
        : testResult.consistency_score;
    }

    let weakWords: string[] = (analytics.weakWords as string[]) ?? [];
    if (testResult.weak_words) {
      const combined = [...weakWords, ...(testResult.weak_words as string[])];
      const wordCounts = new Map<string, number>();
      for (const w of combined) {
        if (typeof w === 'string') {
          wordCounts.set(w, (wordCounts.get(w) ?? 0) + 1);
        }
      }
      weakWords = [...wordCounts.entries()]
        .sort((a, b) => b[1] - a[1])
        .slice(0, 50)
        .map(([w]) => w);
    }

    const last20TestIds: string[] = (analytics.last20TestIds as string[]) ?? [];
    last20TestIds.push(testResult.id as string);
    const updatedLast20 = last20TestIds.slice(-20);

    const wpmTrend: any[] = (analytics.wpmTrend as any[]) ?? [];
    if (testResult.net_wpm != null) {
      wpmTrend.push({ test_id: testResult.id, wpm: testResult.net_wpm, date: new Date().toISOString() });
    }
    const updatedWpmTrend = wpmTrend.slice(-50);

    const accuracyTrend: any[] = (analytics.accuracyTrend as any[]) ?? [];
    if (testResult.accuracy != null) {
      accuracyTrend.push({ test_id: testResult.id, accuracy: testResult.accuracy, date: new Date().toISOString() });
    }
    const updatedAccuracyTrend = accuracyTrend.slice(-50);

    await db
      .update(userAnalytics)
      .set({
        totalTests,
        totalTimeSeconds,
        avgWpm,
        avgAccuracy,
        bestWpm,
        bestAccuracy,
        consistencyScore,
        weakWords,
        last20TestIds: updatedLast20,
        wpmTrend: updatedWpmTrend,
        accuracyTrend: updatedAccuracyTrend,
        updatedAt: new Date(),
      })
      .where(eq(userAnalytics.userId, userId));
  }

  async getLeaderboard(params: {
    scope?: string;
    period?: string;
    limit?: number;
    userId?: string;
    state?: string;
    district?: string;
  }) {
    const limit = params.limit ?? 100;
    let query = db
      .select({
        id: users.id,
        fullName: users.fullName,
        state: users.state,
        district: users.district,
        xp: users.xp,
        level: users.level,
      })
      .from(users)
      .orderBy(desc(users.xp))
      .limit(limit);

    const entries: any[] = [];
    let userRank: number | null = null;

    const result = await query;

    for (let idx = 0; idx < result.length; idx++) {
      const user = result[idx];
      const rank = idx + 1;
      entries.push({
        rank,
        userId: user.id,
        fullName: user.fullName,
        state: user.state,
        district: user.district,
        xp: user.xp,
        level: user.level,
      });
      if (params.userId && user.id === params.userId) {
        userRank = rank;
      }
    }

    return { entries, userRank, totalUsers: result.length };
  }

  async recordErrorPattern(userId: string, patternType: string, patternValue: string) {
    const [existing] = await db
      .select()
      .from(errorPatterns)
      .where(
        and(
          eq(errorPatterns.userId, userId),
          eq(errorPatterns.patternType, patternType),
          eq(errorPatterns.patternValue, patternValue),
        )
      )
      .limit(1);

    if (existing) {
      await db
        .update(errorPatterns)
        .set({
          frequency: existing.frequency + 1,
          lastOccurredAt: new Date(),
        })
        .where(eq(errorPatterns.id, existing.id));
    } else {
      await db.insert(errorPatterns).values({
        userId,
        patternType,
        patternValue,
        frequency: 1,
        lastOccurredAt: new Date(),
      });
    }
  }
}

export const analyticsService = new AnalyticsService();
